packages/db/src/query/compiled-query.ts

```
@@ -69,16 +75,26 @@ export class CompiledQuery<TResults extends object = Record<string, unknown>> {
             .forEach((changes, rawKey) => {
               const { deletes, inserts, value } = changes
               const valueWithKey = { ...value, _key: rawKey }
-              if (inserts && !deletes) {
+
+              // Simple singular insert.
+              if (inserts && deletes === 0) {
                 write({
                   value: valueWithKey,
                   type: `insert`,
                 })
-              } else if (inserts >= deletes) {
+              } else if (
+                // Insert & update(s) (updates are a delete & insert)
+                inserts > deletes ||
+                // Just update(s) but the item is already in the collection (so
+                // was inserted previously).
+                (inserts === deletes &&
+                  collection.has(valueWithKey._key as string | number))
+              ) {
                 write({
                   value: valueWithKey,
                   type: `update`,
                 })
+                // Only delete is left as an option
```

I think the logic here is correct.

It may be good to add an `else` that throws with a "this should never happen" error. Currently, if the logic is wrong, the error is silently eaten but we want to know about it.

I quite like the pattern I have seen others using where they have a `shouldNeverHappen()` function that in dev mode triggers a breakpoint with `debugger`, and in prod throws an error.